### PR TITLE
Recommendations hidden fields

### PR DESCRIPTION
**Pull request by mocca102**
Support hiddenFields parameter for Recommendations requests

### CODE REVIEW

File: AutocompleteClientTests/FW/Logic/Worker/ConstructorIORecommendationsTests.swift [hunk 1]
```diff
@@ -257,4 +257,18 @@ class ConstructorIORecommendationsTests: XCTestCase {
...
+    
+    func testRecommendations_UsingRecommendationsQueryBuilder_AttachesHiddenFields() {
+        let hiddenFields = ["hidden_field_1", "hidden_field_2"]
+        let query = CIORecommendationsQueryBuilder(podID: "item_page_1")
+            .setHiddenFields(hiddenFields)
+            .build()
+
+        let builder = CIOBuilder(expectation: "Calling Recommendations with hidden fields with the same name should have a multiple facet URL query items", builder: http(200))
+        
+        stub(regex("https://ac.cnstrc.com/recommendations/v1/pods/item_page_1?_dt=\(kRegexTimestamp)&c=\(kRegexVersion)&fmt_options%5Bhidden_fields%5D=hidden_field_1&fmt_options%5Bhidden_fields%5D=hidden_field_2&i=\(kRegexClientID)&key=\(kRegexAutocompleteKey)&num_results=5&s=\(kRegexSession)&section=Products"), builder.create())
```
Comment (jjl014) [inline AutocompleteClientTests/FW/Logic/Worker/ConstructorIORecommendationsTests.swift:269]: It looks like we're missing `&\(TestConstants.defaultSegments)` at the end here which is causing the test to fail

**Review by jjl014 — CHANGES_REQUESTED** [review]
Thanks for working on this!

Mostly looking good, but left a couple comments around some of the tests failing.

File: AutocompleteClientTests/FW/Logic/Worker/ConstructorIOIntegrationTests.swift [hunk 2]
```diff
@@ -281,6 +281,25 @@ class ConstructorIOIntegrationTests: XCTestCase {
         self.wait(for: expectation)
     }
 
+    func testRecommendations_WithHiddenFields() {
+        let expectation = XCTestExpectation(description: "Request 204")
+        let hiddenFields = ["testField"]
+        let query = CIORecommendationsQuery(podID: podID, hiddenFields: hiddenFields, section: sectionName)
```
[the remark this answers is not in the document]
Comment (Mudaafi) [reply AutocompleteClientTests/FW/Logic/Worker/ConstructorIOIntegrationTests.swift:287]: Yeah I don't think we have enough behavioral data for the `alternative` strategy but the filtered strategy is working if we really want to confirm

[the remark this answers is not in the document]
Comment (jjl014) [reply AutocompleteClientTests/FW/Logic/Worker/ConstructorIOIntegrationTests.swift:287]: Ooo, that's a good call!

[the remark this answers is not in the document]
Comment (mocca102) [reply AutocompleteClientTests/FW/Logic/Worker/ConstructorIOIntegrationTests.swift]: Thanks for the hints. Fixed here https://github.com/Constructor-io/constructorio-client-swift/pull/223/commits/731e4de30905c7c53803a3b4728e3a26298090ed#diff-02d2d47b304aa89db2d5938ccf20345e778c2eb8c78dcd8a1a67d51bafa2d1eaR288

File: AutocompleteClientTests/FW/Logic/Worker/ConstructorIORecommendationsTests.swift (hunk 1 above)
Comment (mocca102) [reply AutocompleteClientTests/FW/Logic/Worker/ConstructorIORecommendationsTests.swift]: https://github.com/Constructor-io/constructorio-client-swift/pull/223/commits/731e4de30905c7c53803a3b4728e3a26298090ed#diff-a1cb8911026626ff1ea273da2a15eca4e0d0d2050034b918fd4b56f16581f2a0R269

File: AutocompleteClientTests/FW/Logic/Request/RecommendationsQueryRequestBuilderTests.swift [hunk 3]
```diff
@@ -139,4 +139,17 @@ class RecommendationsQueryRequestBuilderTests: XCTestCase {
...
         XCTAssertTrue(url.hasPrefix("\(customBaseURL)/recommendations/v1/pods/\(podID)?"))
     }
+    
+    func testRecommendationsQueryBuilder_WithHiddenFields() {
+        let hiddenFields = ["hiddenField1", "hiddenField2"]
+        let query = CIORecommendationsQuery(podID: self.podID, hiddenFields: hiddenFields)
+        builder.build(trackData: query)
+        let request = builder.getRequest()
+        let url = request.url!.absoluteString
+        XCTAssertTrue(url.hasPrefix("https://ac.cnstrc.com/recommendations/v1/pods/\(podID)?"))
```
[the remark this answers is not in the document]
Comment (mocca102) [reply AutocompleteClientTests/FW/Logic/Request/RecommendationsQueryRequestBuilderTests.swift]: https://github.com/Constructor-io/constructorio-client-swift/pull/223/commits/731e4de30905c7c53803a3b4728e3a26298090ed#diff-61430499c1c50a2ffc8cdc5da72ccfb3023dab9d821f60da68572ccf05aa6515R149

File: AutocompleteClientTests/FW/Logic/Worker/ConstructorIORecommendationsTests.swift [hunk 4]
```diff
@@ -257,4 +257,18 @@ class ConstructorIORecommendationsTests: XCTestCase {
...
         self.wait(for: builder.expectation)
     }
+    
+    func testRecommendations_UsingRecommendationsQueryBuilder_AttachesHiddenFields() {
+        let hiddenFields = ["hidden_field_1", "hidden_field_2"]
+        let query = CIORecommendationsQueryBuilder(podID: "item_page_1")
+            .setHiddenFields(hiddenFields)
+            .build()
+
+        let builder = CIOBuilder(expectation: "Calling Recommendations with hidden fields with the same name should have a multiple facet URL query items", builder: http(200))
```
[the remark this answers is not in the document]
Comment (mocca102) [reply AutocompleteClientTests/FW/Logic/Worker/ConstructorIORecommendationsTests.swift]: https://github.com/Constructor-io/constructorio-client-swift/pull/223/commits/731e4de30905c7c53803a3b4728e3a26298090ed#diff-a1cb8911026626ff1ea273da2a15eca4e0d0d2050034b918fd4b56f16581f2a0R267

**Comment by mocca102** [comment]
@Mudaafi Added parser tests also

**Review by jjl014 — APPROVED** [review]
LGTM! Thank you for working on this! ❤️

**Pull request merged**